Soft deprecate "watchify.args"

zertosh opened the pull request:
Users are unwittingly sharing caches across browserify instances. There is discussion on alternatives in https://github.com/substack/watchify/pull/204.

The solution proposed here is just the basis for a developer education campaign. The idea is to still have `watchify.args`, but not document it. Instead insist that users use `{cache: {}, packageCache: {}}` in the main README, and across the various "recipes" sites (e.g. [gulp](https://github.com/gulpjs/gulp/blob/master/docs/recipes/fast-browserify-builds-with-watchify.md), [gulp-watchify](https://github.com/marcello3d/gulp-watchify/blob/0ff6362f8ad521fb357c8a6563e36e932888bea0/index.js#L17)). This way we don't break existing code.

I'm on the fence about `watchify.defaults` being internal or not.

mattdesl commented:
> The solution proposed here is just the basis for a developer education campaign. The idea is to still have watchify.args, but not document it. Instead insist that users use {cache: {}, packageCache: {}} in the main README, and across the various "recipes" sites (e.g. gulp, gulp-watchify). This way we don't break existing code.

:+1: seems good to me

jmm commented:
My 2 cents: the thing that I like about something like `watchify.args` (or more accurately something like `watchify.defaults()`) is that it provides some insulation from change. If that'll set whatever browserify options are required for whatever version of watchify is used then you could just promote that in all of the docs / recipes and not have to worry about updating them if the required options change, like when you made `fullPaths` optional. And the users also have to concern themselves less than if they set the options explicitly. Like now, how many people who previously set the options explicitly will keep using `fullPaths: true` for how long before they realize they don't need to? Obviously if going this route I would note in the docs that the default options may change and the user should set them explicitly if they want to retain full control, and assume responsibility for making sure required options are set.

zertosh commented:
> Obviously if going this route I would note in the docs that the default options may change and the user should set them explicitly if they want to retain full control, and assume responsibility for making sure required options are set.

I've been persuaded :smile:  How would you explain it though? Should it be the primary-way (i.e. encouraged way) to use `watchify.defaults()`?

zertosh commented:
Should it be called something else? Because `_.defaults` doesn't overwrite properties, but `watchify.defaults` does.

jmm commented:
> Should it be called something else? Because _.defaults doesn't overwrite properties, but watchify.defaults does.

In #204 I suggested `getArgs()` (although I didn't put a lot of thought into the specific name -- I just wanted to float the idea). Perhaps `reqdOpts` or `xtendOpts` would be good?

> I've been persuaded :smile: How would you explain it though? Should it be the primary-way (i.e. encouraged way) to use `watchify.defaults()`?

I still like the idea I floated in #204:

> Offer as a convenience to construct a browserify instance for the user:
```js
// Set `cache` and `packageCache` on the browserify opts
// if `undefined`
var w = watchify({debug: true, ...}, {delay: 600, ...});
```

I'm guessing most people do:

```js
var b = browserify({
  ...
});

var w = watchify(b);

// or

var w = watchify(browserify({
  ...
  
});
```

(As opposed to constructing the browserify instance in one scope and passing it to another scope where it's wrapped with watchify.)

Do you think that's the case?

> How would you explain it though?

Here's a way it could be explained (incorporating multiple unsettled ideas):

<hr />

You MUST pass certain options to the browserify constructor. There are several ways to do it:

1. Pass your browserify options to watchify and let it extend the required options with them and construct the browserify instance for you:

  ```js
  watchify(
    // browserify options that will extend the options
    // required by watchify.
    {debug: true, ...}
  )  
  ```

2. Pass your browserify options to `watchify.reqdOpts()`, which will extend the required options with them:

  ```js
  var w = watchify(browserify(watchify.reqdOpts({
    debug: true,
    ...
  }));
  ```

3. Pass the required options to browserify yourself. Currently they are `{cache: {}, packageCache: {}}`. The required options are subject to change, so if you use this approach check the documentation when changing watchify versions.

<hr />

And then in external documentation ("recipes" sites, etc.) link to the documentation here for the currently required options.

jmm commented:
@zertosh Re: constructing the browserify instance for the user...
> It was like that before v1.0.0 (e.g. v0.10.2). 

Do you have any idea why that was changed?

> The way it is now, watchify behaves like a true plugin (with a leaky cache implementation).

So currently you could actually do this, right?:

```js
browserify({
  cache: {},
  packageCache: {}
})
  .plugin(watchify, {delay: 600, ...})
```

But that's not how it's documented. As long as the usage is going to be documented as:

```js
var b = browserify({ cache: {}, packageCache: {} });
var w = watchify(b);
```
Would it not make sense to take advantage of the ability that provides to set the required options in a more elegant way? Just wondering if there's something I don't understand about why that was changed or potential downsides.

And by the way, I've seen (and maybe you have too) at least one issue on browserify where someone was asking for a way to run plugin functionality before the browserify instance is constructed, to be able to do something like `browserify().plugin(watchify)` and actually make it work (e.g. extend the browserify options before it's constructed).

zertosh commented:
> Do you have any idea why that was changed?

Not really.

> So currently you could actually do this, right?:

Yup. That's actually how I use it personally and at SoundCloud. Real code:

```js
var b = _.once(function() {
  var b_ = require('browserify')({
    entries: './app/main.js',
    detectGlobals: false,
    noparse: ['immutable', 'rsvp', 'underscore', 'classnames'],
    extensions: ['.jsx'],
    debug: process.env.NODE_ENV !== 'production',
    standalone: 'main',
    cache: {}, packageCache: {} // watchify options
  });
  b_.transform(require('./resources/preprocessor'));
  b_.transform(require('envify'));
  b_.plugin(require('dep-case-verify'));
  if (process.env.NODE_ENV === 'production') {
    b_.transform(require('unreachable-branch-transform'));
    b_.plugin(require('bundle-collapser/plugin'));
    b_.plugin(require('bundle-minify'));
  }
  if (isWatching) {
    b_.plugin(require('errorify'));
    b_.plugin(require('watchify'), {delay: 400, ignoreWatch: true});
    // micro-optimization
    b_.listeners('reset').some(function(handler) {
      if (String(handler).indexOf('(delta / 1000)') !== -1) {
        b_.removeListener('reset', handler);
        return true;
      }
    });
  }
  return b_;
});
```

^ Fun facts: You can call `b()` from wherever, and get the browserify instance w/o having to care if it's been created already or not. Also, if you're only building css or watching the server, you don't spend time loading all those requires - lazy loading. Of course, you can still do this via the suggested route, but I like seeing what's a transform and what's a plugin :smile: 

> Would it not make sense to take advantage of the ability that provides to set the required options in a more elegant way? Just wondering if there's something I don't understand about why that was changed or potential downsides.

I had nothing to do with that change, but before it, I think I had to do something like:

```js
var b = (isWatching ? watchify : browserify)(opts);
```

Which is kinda gross. However, doing it the way bundle-collapser does it would be fine too. The only thing that stops me from doing that is that I'd have to defacto rename `index.js`. And then there goes the easy github history. If https://github.com/substack/watchify/pull/204#issuecomment-97630419 doesn't work, then I'll make a PR for that approach and ask substack to weigh in - if he's cool with the git history having that break, then I'm cool with it too.

> at least one issue on browserify where someone was asking for a way to run plugin functionality before the browserify instance is constructed

I don't think that's possible w/o a ton of refactoring. A pipeline *has* to be constructed before you can add transforms/entries/requires. Because of the whole recorder mode thing (which I think is a genius idea).

jmm commented:
> Yup. That's actually how I use it personally and at SoundCloud. Real code:

Oh interesting!

> ^ Fun facts:...

Yeah, seems like a pretty slick setup.

> but before it, I think I had to do something like:
```js
var b = (isWatching ? watchify : browserify)(opts);
```
> Which is kinda gross.

That part actually doesn't bother me. To me the messier part of that would be dealing with whether or not to pass watchify args in the second position. But overall I'm not sure the current handling of the required args makes things neater, e.g.:

```js
var b, opts = {debug: true, ...};
if (isWatching) {
  opts = xtend(opts, watchify.args);
}
b = (isWatching ? watchify : browserify)(opts)
```

> However, doing it the way bundle-collapser does it would be fine too. The only thing that stops me from doing that is that I'd have to defacto rename `index.js`. And then there goes the easy github history. If #204 (comment) doesn't work, then I'll make a PR for that approach and ask substack to weigh in - if he's cool with the git history having that break, then I'm cool with it too.

The easiest thing all around wouldn't be to keep everything as-is but construct a browserify instance if one isn't passed in? It's the same signature (so it doesn't interfere with using it via wrapping or via `.plugin()`) and it's easy enough to detect whether the arg is an instance or an options hash. Wouldn't the bundle-collapser organization (`index.js` and `plugin.js`) mean you'd have to require a different module depending whether you want to wrap an existing browserify instance or let it construct a new one, and still need the same kind of conditional logic that you illustrated?

```js
b = (isWatching ? require('watchify') : browserify)(opts);

// or, instance already exists
if (isWatching) b = require('watchify/plugin')(b);
```

That seems messier to me, unless there's some compelling advantage to it. I may misunderstand something key though.

> I don't think that's possible w/o a ton of refactoring. A pipeline has to be constructed before you can add transforms/entries/requires. Because of the whole recorder mode thing (which I think is a genius idea).

Could be. I didn't give it much thought until yesterday. I would have to really look into it to see if it'd be feasible, but I'd start exploring something like: put the pipeline in a pending state, `nextTick` a bunch of the constructor stuff, stash away or `nextTick` calls that write to the pipeline, see if plugins have a `pre()` method and if so give it a crack at the constructor opts.... But even if it's doable, that's probably what you mean by a ton of refactoring :)

LinusU commented on index.js:
```diff
@@ -4,9 +4,11 @@ var chokidar = require('chokidar');
+    return xtend({cache: {}, packageCache: {}}, opts);
 };
+// DEPRECATED - will be removed in a future version
+module.exports.args = watchify.defaults();
```
How about using the builtin node.js deprecate method? This will allow the use of `--trace-deprecation` to find the offending code.

```javascript
var util = require('util')

Object.defineProperty(module.exports, 'args', {
  get: util.deprecate(watchify.defaults, 'watchify.args: use watchify.default(...) instead')
})
```

This would also solve the problem of shared cache since each `var _ = watchify.args` would get its own object. Example:

```javascript
var a = watchify.args
var b = watchify.args

console.log('is same:', (a === b).toString())
```

LinusU replied:
If you don't want `watchify.args` to return different instance each time:

```javascript
var util = require('util')

var cachedArgs
Object.defineProperty(module.exports, 'args', {
  get: function () {
    if (cachedArgs) return cachedArgs

    cachedArgs = util.deprecate(watchify.defaults, 'watchify.args: use watchify.default(...) instead')()
    return cachedArgs
  }
})
```

zertosh replied:
I'm not a fan of magical getters/setters. I don't think I want to do away with `args` just yet. (notice how old this PR is)

zertosh commented:
I removed the reference to `watchify.args` from the readme, and added a few more examples that should drive the point home (see e8f0217). I don't see these going away, but by not advertising them, hopefully nobody will fall into the trap.
